lib/msf/util/windows_registry_parser.rb

```diff
@@ -0,0 +1,362 @@
+      end
+
+      nil
+    end
+
+    def get_block(offset)
+      RegHbinBlock.read(@hive_data[4096+offset..-1])
+    end
+
+    def compare_hash(magic, hash_rec, key)
+      case magic
+      when LF_MAGIC
+        if hash_rec.key_name.gsub(/(^\x00*)|(\x00*$)/, '') == key[0,4]
+          return hash_rec.offset_nk
+        end
+      when LH_MAGIC
+        if hash_rec.key_name.unpack('<L').first == get_lh_hash(key)
+          return hash_rec.offset_nk
+        end
+      when RI_MAGIC
+        # Special case here, don't know exactly why, an RI pointing to a NK
+        offset = hash_rec.offset_nk
+        nk = get_block(offset)
+        return offset if nk.key_name == key
+      else
+        raise ArgumentError, "Unknow magic: #{magic}"
+      end
+    end
+
+    def get_lh_hash(key)
+      res = 0
+      key.upcase.bytes do |byte|
+        res *= 37
+        res += byte.ord
+      end
+      return res % 0x100000000
+    end
+
+    def get_value_blocks(offset, count)
+      value_list = []
+      res = []
+      count.times do |i|
+        value_list << @hive_data[4096+offset+i*4, 4].unpack('<l').first
+      end
+      value_list.each do |value_offset|
+        if value_offset > 0
+          block = get_block(value_offset)
+          res << block
+        end
+      end
+      return res
+    end
+
+    def get_value_data(record)
+      unless record&.magic == VK_MAGIC
+        raise ArgumentError, "get_value_data: record must be a VK record"
+      end
+      return '' if record.data_len == 0
+      # if DataLen < 5 the value itself is stored in the Offset field
+      return record.offset_data.to_binary_s if record.data_len < 0
+      return self.get_data(record.offset_data, record.data_len + 4)
+    end
+
+    def get_data(offset, count)
+      @hive_data[4096+offset, count][4..-1]
+    end
+
+    def enum_key(parent_key)
+      unless parent_key&.data&.magic == NK_MAGIC
+        raise ArgumentError, "enum_key: parent key must be a NK record"
+      end
+      block = get_block(parent_key.data.offset_sub_key_lf)
+      records = []
+      if block.data.magic == RI_MAGIC
+        # ri points to lf/lh records, so we consolidate the hash records in the main records array
+        block.data.hash_records.each do |hash_record|
+          record = get_block(hash_record.offset_nk)
+          records.concat(record.data.hash_records)
+        end
+      else
+        records.concat(block.data.hash_records)
+      end
+
+      records.map do |reg_hash|
+        nk = get_block(reg_hash.offset_nk)
+        nk.data.key_name.to_s.b
+      end
+    end
+
+    def enum_values(key)
+      unless key&.data&.magic == NK_MAGIC
+        raise ArgumentError, "enum_values: key must be a NK record"
+      end
+      res = []
+      value_list = get_value_blocks(key.data.offset_value_list, key.data.num_values + 1)
+      value_list.each do |value|
+        if value.data.flag > 0
+          res << value.data.name
+        else
+          res << 'default'.b
```

Comment:
Should this instead be either `nil` or `:default` instead of a string to differentiate between a default value and a literal value of `"default"`?

Reply:
Absolutely, I will set the default value to `nil`.